### PR TITLE
Added Explanatory docs for VPG

### PR DESCRIPTION
**Pull request by Devanshu24**
Sorry for the delay, I had been prolonging it for no reason
This is just the most minimal implementation, please let me know of any additions that ﻿need to be made

### CODE REVIEW

**Comment by sampreet-arthi** [comment]
You can add a few more details here. Maybe include some hyperparams like rollout size for example. Also add a use case for using it with Atari envs.

File: docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Vanilla Policy Gradient (VPG)
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md:1]: could you rename to file to `using_vpg.md`

In general its better to not include spaces in filesnames

File: docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+env = VectorEnv("CartPole-v1")
+agent = VPG('mlp', env)
+trainer = OnPolicyTrainer(agent, env, epochs=200)
+trainer.train()
+```
+
+Source Codes: \
+`VPG` class  [here](https://github.com/SforAiDl/genrl/blob/9ce9716b18aab490e880c99cd530965db6bc4ba9/genrl/deep/agents/vpg/vpg.py)\
+`OnPolicyTrainer` class [here](https://github.com/SforAiDl/genrl/blob/9ce9716b18aab490e880c99cd530965db6bc4ba9/genrl/deep/common/trainer.py#L406)\
+`VectorEnv` method[here](https://github.com/SforAiDl/genrl/blob/9ce9716b18aab490e880c99cd530965db6bc4ba9/genrl/environments/suite.py#L16)
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md:21]: Is this necessary? I think it would be better to point to the individual docs for these. As has been done in the Using A2C

File: docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md (hunk 1 above)
Comment (Devanshu24) [reply docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md:1]: Sure

**Comment by Devanshu24** [comment]
> We need to introduce more variation to these. Otherwise all these are just similar to what's already present in the readme. These are the things you could do:

Completely agreed

I'll make the changes!

File: docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md (hunk 2 above)
Comment (Devanshu24) [reply docs/source/usage/tutorials/Vanilla Policy Gradient (VPG).md:21]: The A2C one doesn't work
Should I just add [this](https://genrl.readthedocs.io/en/latest/api/algorithms/genrl.deep.agents.vpg.html) instead

**Comment by Sharad24** [comment]
Any progress?

**Comment by Devanshu24** [comment]
Yes, will push them today

**Comment by Devanshu24** [comment]
There might be some conceptual mistakes (I tried to not have them)
But if there are any changes needed conceptual or otherwise please do tell

PS: I haven't added the Atari version as it will be the same code as other algos, if needed I'll do so

**Comment by sampreet-arthi** [comment]
> PS: I haven't added the Atari version as it will be the same code as other algos, if needed I'll do so

Could you do this anyway? It would be more complete. The tutorials should be as clear as possible.

File: docs/source/usage/tutorials/using_vpg.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+If you wanted to explore Policy Gradient algorithms in RL, there is a high chance you would've heard of PPO, DDPG, etc. but understanding them can be tricky if you're just starting.
+
+VPG is arguably one of the easiest to understand policy gradient algorithms while still performing to a good enough level
+
+Let's understand policy gradient at a high level, unlike the classical algorithms like Q-Learning, Monte Carlo where you try to optimise the outputs of the action-value function of the agent which are then used to determine the optimal policy. In policy gradient, as one would like to say we go directly for the kill shot, basically we optimise the thing we want to use at the end, i.e. the Policy.
+
+So that explains the "Policy" part of Policy Gradient, so what about "Gradient", so gradient comes from the fact that we try to optimise the policy by gradient ascent (unlike the popular gradient descent, here we want to increase the values, hence ascent). So that explains the name, but how does it even work.
+
+For that, have a look at the following Psuedo Code (source: [OpenAI](spinningup.openai.com))
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/using_vpg.md:11]: You can add https://spinningup.openai.com/en/latest/spinningup/rl_intro3.html as a source for more detailed derivation also.

File: docs/source/usage/tutorials/using_vpg.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+from genrl.environments import VectorEnv
+
+env = VectorEnv("CartPole-v1")
+agent = VPG('mlp', env)
+trainer = OnPolicyTrainer(agent, env, epochs=200)
+trainer.train()
+```
+
+This will run a VPG agent `agent` which will interact with the `CartPole-v1` environment \
+Let's understand the output on running this(your individual values may differ),
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/using_vpg.md:32]:
```suggestion
Let's understand the output on running this (your individual values may differ),
```

File: docs/source/usage/tutorials/using_vpg.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+389120           190              23.4676          177.316
+```
+
+`timestep`: It is basically the units of time the agent has interacted with the environment since the start of training\
+`Episode`: It is one complete rollout of the agent, to put it simply it is one complete run until the agent ends up winning or losing\
+`loss`: The loss encountered in that episode\
+`mean_reward`: The mean reward accumulated in that episode\
+
+Now if you look closely the agent will not converge to the max reward even if you increase the epochs to say 5000, it is because that during training the agent is behaving to a stochastic policy\
+So even if the agent has figured out the optimal policy it is not taking the most optimal action at every step there is an inherent stochasticity to it\
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/using_vpg.md:64]: Could you explain stochaticity a bit more here. Something like the agent's policy function outputs a probability distribution over the actions which we sample from for the action.

File: docs/source/usage/tutorials/using_vpg.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+from genrl.deep.common import OnPolicyTrainer
+from genrl.environments import VectorEnv
+
+env = VectorEnv("CartPole-v1")
+agent = VPG('mlp', env)
+trainer = OnPolicyTrainer(agent, env, epochs=200)
+trainer.train()
+```
+
+This will run a VPG agent `agent` which will interact with the `CartPole-v1` environment \
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/using_vpg.md:31]: You can link to the openai gym website with a one line explanaiton about environments since beginners maybe not be aware of them.

**Comment by threewisemonkeys-as** [comment]
Added few minor suggestions.

**Comment by sampreet-arthi** [comment]
Also, @Darshan-ko could you take note of these suggestions too in the bandit tutorials and also modify the A2C tutorial you had pushed earlier? There's more clarity regarding these now ig.

File: docs/source/usage/tutorials/using_vpg.md (hunk 3 above)
Comment (Devanshu24) [reply docs/source/usage/tutorials/using_vpg.md:11]: Sure

File: docs/source/usage/tutorials/using_vpg.md (hunk 5 above)
Comment (Devanshu24) [reply docs/source/usage/tutorials/using_vpg.md:64]: Okay, even I felt this was not that great. Will do!

File: docs/source/usage/tutorials/using_vpg.md (hunk 6 above)
Comment (Devanshu24) [reply docs/source/usage/tutorials/using_vpg.md:31]: Sure

**Comment by Devanshu24** [comment]
On a side note:
What is more preferred
1. I keep pushing commits again and again
2. I make the changes and keep doing a rebase (squash/fixup) to make just a few commits that are more meaningful

I know that at the end you just squash and merge so it won't impact the master branch.
But still any preference for the PR per se?

**Comment by sampreet-arthi** [comment]
Commit less frequently and only for relevant grouped changes. Push only when you get reviews or want to check CI.

**Comment by Devanshu24** [comment]
I have made all the said changes, if anything is off or needs to be added please do tell
